pennylane/devices/qubit/simulate.py

```diff
@@ -316,6 +347,334 @@ def simulate_partial(k):
+            If no value is provided, a default RNG will be used.
+        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. This is
+            the key to the JAX pseudo random number generator. If None, a random key will be
+            generated. Only for simulation using JAX.
+        debugger (_Debugger): The debugger to use
+        interface (str): The machine learning interface to create the initial state with
+        initial_state (TensorLike): Initial state vector
+        mcm_active (dict): Mid-circuit measurement values or all parent circuits of ``circuit``
+        mcm_samples (dict): Mid-circuit measurement samples or all parent circuits of ``circuit``
+
+    Returns:
+        tuple(TensorLike): The results of the simulation
+    """
+    interface = execution_kwargs.get("interface", None)
+
+    samples_present = any(isinstance(mp, SampleMP) for mp in circuit.measurements)
+    postselect_present = any(
+        op.postselect is not None for op in circuit.operations if isinstance(op, MidMeasureMP)
+    )
+    if postselect_present and samples_present and circuit.batch_size is not None:
+        raise ValueError(
+            "Returning qml.sample is not supported when postselecting mid-circuit "
+            "measurements with broadcasting"
+        )
+
+    #########################
+    # shot vector treatment #
+    #########################
+    if circuit.shots.has_partitioned_shots:
+        results = []
+        for s in circuit.shots:
+            aux_circuit = qml.tape.QuantumScript(
+                circuit.operations,
+                circuit.measurements,
+                shots=qml.measurements.Shots(s),
+                trainable_params=circuit.trainable_params,
+            )
+            results.append(simulate_tree_mcm(aux_circuit, debugger, **execution_kwargs))
+        return tuple(results)
+
+    #######################
+    # main implementation #
+    #######################
+
+    def init_dict(d):
+        return {} if d is None else d
+
+    mcm_active = init_dict(mcm_active)
+    mcm_samples = init_dict(mcm_samples)
+
+    circuit_base, circuit_next, op = circuit_up_to_first_mcm(circuit)
+    # we need to make sure the state is the all-wire state
+    initial_state = prep_initial_state(circuit_base, interface, initial_state, circuit.wires)
+    state, is_state_batched = get_final_state(
+        circuit_base,
+        debugger=debugger,
+        initial_state=initial_state,
+        mid_measurements=mcm_active,
+        **execution_kwargs,
+    )
+    measurements = measure_final_state(
+        circuit_base, state, is_state_batched, initial_state=initial_state, **execution_kwargs
+    )
+
+    # Simply return measurements when ``circuit_base`` does not have an MCM
+    if circuit_next is None:
+        return measurements
+
+    # For 1-shot measurements as 1-D arrays
+    samples = measurements.reshape((-1)) if measurements.ndim == 0 else measurements
```

Comment:
Confused about the -1 value (is it the same as if you put `[1]`?)

Reply:
`reshape((-1))` is the same as `ravel()`.

Reply:
Would that be more expressive then? Looking at the docs for reshape, it says the provided shape is allowed to have one unknown dimension (indicated with a `-1`), so while it does appear to turn the scalar into a 1-D array (because the shape contains one dimension), it is a bit less intuitive.

Actually, this function would be even more expressive I think and also get rid of the if statement:
https://numpy.org/doc/stable/reference/generated/numpy.atleast_1d.html

Reply:
Happy to change to `ravel` if you prefer it.

Reply:
Actually I think `atleast_1d` is better, does that correspond to your intention?

Reply:
No, it's not `atleast_1d`, it's `exactly_1d`.

Reply:
Doesnt your code do this?
```
if ndim is 0:
 promote to 1d
else:
 leave as is 
```

`atleast_1d` does exactly that

Reply:
Right, by bad. I'll change it.